src/weak.rs

```diff
@@ -0,0 +1,59 @@
+use core::{
+    ffi::c_void,
+    ptr::NonNull,
+    sync::atomic::{fence, AtomicPtr, Ordering},
+};
+
+// A "weak" binding to a C function that may or may not be present at runtime.
+// Used for supporting newer OS features while still building on older systems.
+// Based off of the DlsymWeak struct in libstd:
+// https://github.com/rust-lang/rust/blob/1.61.0/library/std/src/sys/unix/weak.rs#L84
+// except that the caller must manually cast self.ptr() to a function pointer.
+pub struct Weak {
+    addr: AtomicPtr<c_void>,
+    link_fn: fn() -> *mut c_void,
+}
+
+impl Weak {
+    // A non-null pointer value which indicates we are uninitialized. This
+    // constant should ideally not be a valid address of a function pointer.
+    // However, if by chance libc::dlsym does return UNINIT, there will not
```

Comment:
// However, if by chance link_fn does return UNINIT, there will not
    // be undefined behavior. link_fn will just be called each time ptr()